superdesk/editor_utils.py

```diff
@@ -401,29 +463,154 @@ def __init__(self, item, field='body_html'):
         self.field = field
         data = item.get('fields_meta', {}).get(field, {})
         if not data:
-            self.content_state = {}
+            self._create_state_from_html(item.get(field), html)
         else:
             # if the field exist, draftjsState must exist too
             self.content_state = data['draftjsState'][0]
         self.blocks = BlockSequence(self)
-
         self.html_exporter = DraftJSHTMLExporter(self)
 
+    def _create_state_from_html(self, value=None, html=True):
+        self.content_state = {
+            'blocks': [],
+            'entityMap': {},
+        }
+
+        if not value:
+            return
+
+        def create_entity(type, data, mutability='MUTABLE'):
+            key = len(self.content_state['entityMap'].keys())
+            self.content_state['entityMap'][str(key)] = {
+                'type': type,
+                'mutability': mutability,
+                'data': data,
+            }
+            return key
+
+        def create_atomic_block(type, data):
+            block = self.create_block('atomic', text=" ").data
+            entity_key = create_entity(type, data)
+            block['entityRanges'] = [{'offset': 0, 'length': 1, 'key': entity_key}]
+            block['inlineStyleRanges'] = []
+            self.content_state['blocks'].append(block)
+
+        if html:
+            root = parse_html(value, 'html')
+            for i, elem in enumerate(root):
+                try:
+                    block_type = next((key for key, val in BLOCK_MAP.items() if val == elem.tag))
+                    depth = 0
+                except StopIteration:
+                    block_type = None
+                    depth = 0
+                    if elem.tag == 'figure' and i > 0 and root[i - 1].text:
+                        m = re.search(r'<!-- EMBED START Image {id: "(.*)"}', str(root[i - 1]).strip())
+                        if m and self.item.get('associations') and self.item['associations'].get(m.group(1)):
+                            media = self.item['associations'][m.group(1)]
+                            create_atomic_block('MEDIA', {'media': media})
+                        continue
+                    elif elem.tag in ('ul', 'ol'):
+                        pass
+                    elif elem.text and '<!-- EMBED' in str(elem):
+                        continue
+                    elif elem.tag == 'table':
+                        data = {'numCols': 0, 'numRows': 0, 'withHeader': False, 'cells': {}}
```

Comment:
Initially I was thinking about creating specialised block classes with high level method, for instance there would be here a `TableBlock` with methods like `addRow` or `addCol`, that would avoid the manipulation of data inside `Editor3Content`. The goal is to let the dev modify internal data if necessary, but to have high level methods to make life easy as much as possible.

@petrjasek @ride90 what do you think?

Reply:
Sure, that makes sense, but I would not spend time with it until manipulations with tables are required.

Reply:
indeed, let's keep it for a later time